server.go

```diff
@@ -17,18 +18,41 @@ import (
 
 // The server handles creating connections and listening on various ports.
 type Server struct {
-	n              *internal.Node
-	listeners      map[string]*internal.SSHListener
-	listen_address string
+	n         *internal.Node
+	listeners map[string]*internal.SSHListener
+	// This contains the addresses of the nodes that we are currently connecting
+	// to, so we don't try to connect to them twice at the same time.
+	currently_connecting map[string]bool
```

Comment:
Hmmm. Okay, you're saying the problem isn't duplicate connections, it's a race with the same connection multiple times. Can we put that in the auto discovery server rather than here?